MAVLink forwarding via UDP

dakejahl opened the pull request:
Added mavlink forwarding as an option to UDP connections. Enabling this setting will cause every mavlink packet received by QGC to be forwarded on the link, with the exception of packets originating from said link.

![image](https://user-images.githubusercontent.com/37091262/82407029-5ba18180-9a25-11ea-919d-d3cf1bc70e4c.png)

Open to any and all suggestions/feedback/criticism/etc. Obviously there are a couple of different ways in which to add this feature, the implementation here was chosen because it was easiest.

dogmaphobic commented on src/comm/LinkManager.cc:
```diff
@@ -1011,6 +1016,21 @@ void LinkManager::_freeMavlinkChannel(int channel)
+        if (forwardMavlink) {
+            qCDebug(LinkManagerLog) << "Forwarding mavlink packet";
+            uint8_t buffer[MAVLINK_MAX_PACKET_LEN];
+            int len = mavlink_msg_to_send_buffer(buffer, &message);
```
This will mess up the message counter. Messages may come from different component IDs and each uses their own counter. Message forwarding must always forward the raw, original message. Not a re-encoded one.

dakejahl replied:
Okay I moved the logic, can you let me know if this is correct? I am also wondering if I need to do anything about the "listening port". I'm not entirely sure how sending on UDP works... I thought there was a separate port for sending and receiving.

dakejahl commented:
@DonLakeFlyer could you review this and let me know if this implementation makes sense? I am a bit confused myself with UDP, there is the "Listening port", but is this the same port on which data is sent? I was under the impression UDP connections were unidirectional. Forgive my ignorance, I am more of an embedded guy.

Antiheavy commented:
should there be a default (or detected) IP and port displayed for this?  How else will someone on the network know where to look for the MAVlink data being re-broadcast from?

dakejahl commented:
Tested successfully using mavlink-router example:
 `python3 examples/receiver.py localhost:14445 1`. 

I had to make a minor adjustment to the python script, so that it first checks for 
`if not hasattr('msg', 'target_system'):`
as QGC heartbeats do not define a target_system attribute. 

When I connected a Pixhawk, it immediately begins dumping the recieved mavlink to the output
```
Message from 1: VFR_HUD {airspeed : 0.0, groundspeed : 0.02274211309850216, heading : 237, throttle : 0, alt : 1276.229736328125, climb : 0.03378766030073166}
Message from 1: ATTITUDE {time_boot_ms : 17602, roll : -0.01602204702794552, pitch : 0.020186109468340874, yaw : -2.1466565132141113, rollspeed : 0.010613268241286278, pitchspeed : -0.00391938304528594, yawspeed : -0.0013091671280562878}
Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17602, q1 : -0.4772334396839142, q2 : -0.0050466363318264484, q3 : -0.011855791322886944, q4 : 0.8786820769309998, rollspeed : 0.010613268241286278, pitchspeed : -0.00391938304528594, yawspeed : -0.0013091671280562878, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
Message from 1: HIGHRES_IMU {time_usec : 17607690, xacc : 0.20465031266212463, yacc : 0.18606188893318176, zacc : -9.777864456176758, xgyro : 0.0009166942909359932, ygyro : 0.00029493076726794243, zgyro : -0.0009300763485953212, xmag : -0.2176124006509781, ymag : 0.20348277688026428, zmag : 0.5183080434799194, abs_pressure : 86907.9921875, diff_pressure : 0.0, pressure_alt : 1276.0421142578125, temperature : 19.219999313354492, fields_updated : 7167, id : 0}
Message from 1: ACTUATOR_CONTROL_TARGET {time_usec : 17626378, group_mlx : 0, controls : [0.013215913437306881, -0.017789743840694427, 0.00038756802678108215, 0.0, 0.0, 0.0, 0.0, -1.0]}
Message from 1: ATTITUDE {time_boot_ms : 17622, roll : -0.01601557992398739, pitch : 0.02017391286790371, yaw : -2.1466526985168457, rollspeed : 0.009854053147137165, pitchspeed : -0.0035595863591879606, yawspeed : -0.0023903457913547754}
Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17622, q1 : -0.4772351086139679, q2 : -0.0050427974201738834, q3 : -0.011850050650537014, q4 : 0.8786812424659729, rollspeed : 0.009854053147137165, pitchspeed : -0.0035595863591879606, yawspeed : -0.0023903457913547754, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
Message from 1: SCALED_IMU {time_boot_ms : 17622, xacc : 16, yacc : -11, zacc : -988, xgyro : -10, ygyro : 1, zgyro : 0, xmag : -213, ymag : 218, zmag : 518, temperature : 0}
Message from 1: SCALED_IMU2 {time_boot_ms : 17623, xacc : -15, yacc : -22, zacc : -986, xgyro : 4, ygyro : 4, zgyro : 0, xmag : 0, ymag : 0, zmag : 0, temperature : 0}
Message from 1: SERVO_OUTPUT_RAW {time_usec : 17623436, port : 0, servo1_raw : 0, servo2_raw : 0, servo3_raw : 0, servo4_raw : 0, servo5_raw : 0, servo6_raw : 0, servo7_raw : 0, servo8_raw : 0, servo9_raw : 0, servo10_raw : 0, servo11_raw : 0, servo12_raw : 0, servo13_raw : 0, servo14_raw : 0, servo15_raw : 0, servo16_raw : 0}
Message from 1: SERVO_OUTPUT_RAW {time_usec : 17625282, port : 1, servo1_raw : 65535, servo2_raw : 65535, servo3_raw : 65535, servo4_raw : 65535, servo5_raw : 4096, servo6_raw : 4096, servo7_raw : 65535, servo8_raw : 65535, servo9_raw : 0, servo10_raw : 0, servo11_raw : 0, servo12_raw : 0, servo13_raw : 0, servo14_raw : 0, servo15_raw : 0, servo16_raw : 0}
Message from 1: HIGHRES_IMU {time_usec : 17627648, xacc : 0.18951965868473053, yacc : 0.15902596712112427, zacc : -9.816869735717773, xgyro : 0.00018552877008914948, ygyro : -0.0017596527468413115, zgyro : -0.001216062344610691, xmag : -0.21456921100616455, ymag : 0.20331481099128723, zmag : 0.5242573618888855, abs_pressure : 86907.0, diff_pressure : 0.0, pressure_alt : 1276.134765625, temperature : 19.219999313354492, fields_updated : 7167, id : 0}
Message from 1: LOCAL_POSITION_NED {time_boot_ms : 17627, x : 0.0, y : 0.0, z : -0.3795027732849121, vx : 0.0208196509629488, vy : -0.009823454543948174, vz : -0.03339726850390434}
Message from 1: ODOMETRY {time_usec : 17627648, frame_id : 18, child_frame_id : 12, x : 0.0, y : 0.0, z : -0.3795027732849121, q : [-0.4772350788116455, -0.0050383200868964195, -0.011848811991512775, 0.8786814212799072], vx : -0.0024241870269179344, vy : 0.02334427461028099, vz : -0.033083293586969376, rollspeed : 0.009519649669528008, pitchspeed : -0.0025424398481845856, yawspeed : -0.0010398122249171138, pose_covariance : [0.08489767462015152, 0.0, 0.0, 0.0, 0.0, 0.0, 0.08489584177732468, 0.0, 0.0, 0.0, 0.0, 0.05443938821554184, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], velocity_covariance : [0.0670827329158783, 0.0, 0.0, 0.0, 0.0, 0.0, 0.06707992404699326, 0.0, 0.0, 0.0, 0.0, 0.009763744659721851, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], reset_counter : 0, estimator_type : 0}
Message from 1: ALTITUDE {time_usec : 17641953, altitude_monotonic : 1275.947021484375, altitude_amsl : 1275.947021484375, altitude_local : 0.37953153252601624, altitude_relative : 0.37953153252601624, altitude_terrain : 0.2677607238292694, bottom_clearance : 0.11177080869674683}
Message from 1: TIMESYNC {tc1 : 0, ts1 : 17642043000}
Message from 1: ATTITUDE {time_boot_ms : 17642, roll : -0.016015376895666122, pitch : 0.02012961357831955, yaw : -2.146639108657837, rollspeed : 0.010402746498584747, pitchspeed : -0.004715935327112675, yawspeed : -0.0019618477672338486}
Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17642, q1 : -0.47724103927612305, q2 : -0.005023301113396883, q3 : -0.01183942798525095, q4 : 0.8786783218383789, rollspeed : 0.010402746498584747, pitchspeed : -0.004715935327112675, yawspeed : -0.0019618477672338486, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
```

dakejahl commented:
Notice the port on which the receiver connects is the "Listening Port" as defined in the UDP Link Configuration Settings
![image](https://user-images.githubusercontent.com/37091262/82764236-89424e00-9dca-11ea-8d50-e100f0cafcc7.png)

dakejahl commented:
Should "Listening Port" be changed to "Port" to reflect the fact that we are both listening on this port and forwarding mavlink on this port? @DonLakeFlyer

DonLakeFlyer commented:
The concept of creating another QGC link to forward to seems odd and won't really work in the case where there is more than one link (multiple vehicle(s) or multiple links on a single vehicle). You can't forward data from more than one link (or mavlink channel) to a single mavlink "channel" if you are just forwarding bytes (and not reconstituting messages which is how telemetry log files work for example). There will be no way to decode that data correctly. Each "QGC Link" which internally equates to a "mavlink channel" can only be forwarded to a single outgoing UDP port associated with that link.

If you want to forward bytes then I think there should just be a UDP Port specified to forward to when you turn on forwarding as a link setting. That said I tend to get lost with the whole UDP Host versus client thing with respect to how to do that correctly. You don't really want a full UDP link in QGC architecture since that is bi-directional and internally is really two UDP ports.

DonLakeFlyer commented:
I think you send it out the UDP port like a broadcast like SITL does. Not sure if that's helpful or not.

DonLakeFlyer commented:
This is also problematic in that to forward you need to create a link. Which means auto-connect stuff won't work. I think the solution here may be to take the telemetry log approach which is to reconstitute messages and then jam them out a UDP port. That would then only require a global "turn on forwarding" setting and a UDP port I think. @dogmaphobic Would know more since he doesn't get lost with UDP like I do.

It's not as "pure" as byte forwarding. But it will allow for simplicity of dealing with multiple links/vehicle and auto-connect.

dakejahl commented:
>You can't forward data from more than one link (or mavlink channel) to a single mavlink "channel"

and 

>Each "QGC Link" which internally equates to a "mavlink channel" can only be forwarded to a single outgoing UDP port associated with that link

I think this is what we want anyways right? Simple 1 to 1 forwarding. @Antiheavy 

>

dakejahl commented:
>This is also problematic in that to forward you need to create a manual link. Which means auto-connect stuff won't work.

Could you elaborate? I have auto-connect turned on for my sik radio and my pixhawk. Having a manual UDP link turned on with forwarding enabled does not affect QGCs ability to auto-connect in either of these instances.

dakejahl commented:
> You can't forward data from more than one link (or mavlink channel) to a single mavlink "channel" 

I didn't understand this correctly the first time I read it. I see now. The bytes between multiple mavlink streams would get mixed and obviously the protocol expects a pack of serialized data for a "message" to be decoded correctly. :+1: 

>reconstitute messages and then jam them out a UDP port

@dogmaphobic mentioned a concern with this
>Message forwarding must always forward the raw, original message. Not a re-encoded one.

DonLakeFlyer commented:
> mentioned a concern with this

It depends on what you are trying to accomplish. It you want the full stream with byte level fidelity of what you got in including trashed packets to come out of QGC on UDP then that statement is correct. If all you need is correct fully formed mavlink messages to come out of QGC, then it's fine.

The interesting thing with respect to using the same conceptual mechanism as telemetry log saving is that Gus' comments on message sequence numbers is correct in that they will get screwed up if you have messages from multiple vehicles getting jammed into a single link using a single mavlink channel. Which basically means that telemetry log playback with multiple vehicles should show incorrect packet loss due to sequence number problems. But since that for that use case that isn't what you are interested in it works fine. I would think the usage would be similar for forwarding. You're not trying to detect packet loss on the forwarded UDP stream. You just want mavlink messages to show up there.

DonLakeFlyer commented:
> Could you elaborate?

Sorry I was jumping ahead to another possible solution. What you say is correct. But given the fact that the way it is implemented now doesn't work. I was jumping ahead to a solution which requires you to specify on a link by link basis whether you want to forward that links traffic to a specific UDP port. That would require creation of a manual link so you can turn it on for that link.

DonLakeFlyer commented:
So after all that I think the way to do this is:
* Global setting to forward mavlink to a specified UDP port on the mavlink settings page
* Reconstitute the messages out the specified UDP port like telemetry logging does it in Mavlinkprotocol.cc
* You don't need to make a strange QGC UDP link in comm link settings

If you want technically "pure" forwarding then use something else like mavlink router.

Antiheavy commented:
> Global setting to forward mavlink to a specified UDP port on the mavlink settings page

this seems a more intuitive place for this.

dakejahl commented:
Okay I updated this PR based on the feedback from @DonLakeFlyer. 

Mavlink settings page:
![image](https://user-images.githubusercontent.com/37091262/82864553-03b1c180-9ee2-11ea-8525-8ce338d2d176.png)

And when forwarding is enabled
![image](https://user-images.githubusercontent.com/37091262/82864578-0f04ed00-9ee2-11ea-90df-3692bdeadfb6.png)

I added the note about restarting the application as I was unsure if we want to allow dynamically updating the UDP link if the host name changes. I added a TODO comment right where I think it should be happening. Let me know if you think this is worth adding.

DonLakeFlyer commented:
Comments:
* In thinking through this I don't think it's safe to use a regular QGC link for the udp forward link. By creating a regular QGC link you create a bi-directional link. Traffic in the reverse direction on that link will be processed if for some reason whatever is on the other side decides to send through it. And then strange things will happen. I think it should just be it's own custom UDP thing which only support write outbound traffic. Using a standard QGC link also causes that link to show up in comm links which is additional oddness. 
* I think both enable and host name setting should set "qgcRebootRequired" which will automatically warn about reboots. For the case of turning it on for the first time I may not be completely true but simpler to use the standard support for this kind of warning.
* The settings UI additions needs to respect to visible property for the settings so custom build can override and hide these if wanted.

dakejahl commented:
>I think it should just be it's own custom UDP thing which only support write outbound traffic

It seems like the easiest way to achieve this would be to add a field to the `UDPLink` class that disables reading, by simply not emitting `bytesReceived`

>Using a standard QGC link also causes that link to show up in comm links which is additional oddness.

Are you talking about the comm links menu page? Because it does not show up there. It is only added to the list of comm links (`_sharedLinks`) as managed by the LinkManager. I could store the link as as a separate member variable `LinkInterface*`, but doing so would require code duplication for creating/managing the link and I didn't see a drawback to doing it this way.

>The settings UI additions needs to respect to visible property for the settings so custom build can override and hide these if wanted.

Could you elaborate? Do I need to do something like:
```
visible:    QGroundControl.settingsManager.appSettings.forwardMavlink.rawValue &&
               parent.visible
```
I don't see any other items in MavlinkSettings.qml that seem to follow any convention for the `visible` property

DonLakeFlyer commented:
> I don't see any other items in MavlinkSettings.qml that seem to follow any convention for the `visible` property

That's because mavlink settings is ages old code which predates the visible property on a setting. Hence it is all incorrect. If you look at general settings you'll see it all over the place.

DonLakeFlyer commented:
> Because it does not show up there.

Ah, ok then. An internal flag for single direction traffic should be ok then.

dakejahl commented:
Alright I added a flag to set the UDPConfig to `transmitOnly`. This sets the sockets receive buffer to zero such that no data is ever received (solution per StackOverflow)

DonLakeFlyer commented on src/comm/UDPLink.cc:
```diff
@@ -293,10 +293,18 @@ bool UDPLink::_hardwareConnect()
 #ifdef __mobile__
         _socket->setSocketOption(QAbstractSocket::SendBufferSizeSocketOption,     64 * 1024);
-        _socket->setSocketOption(QAbstractSocket::ReceiveBufferSizeSocketOption, 128 * 1024);
+        if (_udpConfig->isTransmitOnly()) {
```
The duplicated code on both sides of this idef and inside is getting larger and larger. Too much duplication leads to bug farms down the road when people screw up forgetting to make changes in multiple places. Something like this would be way better while you are in there:
```
#ifdef __mobile__
        int bufferSizeMultiplier = 1;
#else
        int bufferSizeMultiplier = 4;
#endif
        int receiveBufferSize = _udpConfig->isTransmitOnly() ? 0 : 512 * 1024;
        _socket->setSocketOption(QAbstractSocket::SendBufferSizeSocketOption,     bufferSizeMultiplier * 64 * 1024);
        _socket->setSocketOption(QAbstractSocket::ReceiveBufferSizeSocketOption, bufferSizeMultiplier * receiveBufferSize);
```

DonLakeFlyer commented on src/ui/preferences/MavlinkSettings.qml:
```diff
@@ -153,6 +153,38 @@ Rectangle {
+                            width:              _labelWidth
+                            anchors.baseline:   mavlinkForwardingHostNameField.baseline
+                            visible:            QGroundControl.settingsManager.appSettings.forwardMavlink.rawValue &&
+                                                QGroundControl.settingsManager.appSettings.forwardMavlink.visible
```
Should be `QGroundControl.settingsManager.appSettings. QGroundControl.settingsManager.appSettings.forwardMavlinkHostName.visible`. Same thing on the FactTextField below.

DonLakeFlyer replied:
Also most settings which are of the form enabled setting + value setting disable the secondary setting ui when the enable part is not enabled as opposed to hiding. This provides better usability to the user to understand what the setting is for and what will happen if they enable. As opposed to magic extra stuff that shows up when they check something.

dakejahl replied:
Alright I updated this. I made the host name text field always visible but only enabled dependent on the value of the forwardMavlink setting. The visible fields for all the new items are just dependent on the visible property of the forwardMavlink setting, such that if the forwardMavlink setting is disabled everything will go away.

DonLakeFlyer replied:
"The visible fields for all the new items are just dependent on the visible property of the forwardMavlink setting"

That isn't correct., Each field visibility should be based on the visible property for the specific fact it represents.

dakejahl replied:
fixed

DonLakeFlyer commented:
@dakejahl This should be all good to go after this last change, right?

dakejahl commented:
Yes, it was tested successfully by 3 different parties so far. If there's anything we need to change about this in the future just let me know.

ryanjAA commented:
@dakejahl any reason to not let this receive as well (or a toggle for that)? So you could have another GCS instance exerting control, etc if you wanted.

dakejahl commented:
It's been a while so I can't remember the exact details, but I think it's just more involved and I didn't require bidirectional forwarding at the time so I didn't include it in this PR.

yeshangjun commented:
> Tested successfully using mavlink-router example: `python3 examples/receiver.py localhost:14445 1`.
> 
> I had to make a minor adjustment to the python script, so that it first checks for `if not hasattr('msg', 'target_system'):` as QGC heartbeats do not define a target_system attribute.
> 
> When I connected a Pixhawk, it immediately begins dumping the recieved mavlink to the output
> 
> ```
> Message from 1: VFR_HUD {airspeed : 0.0, groundspeed : 0.02274211309850216, heading : 237, throttle : 0, alt : 1276.229736328125, climb : 0.03378766030073166}
> Message from 1: ATTITUDE {time_boot_ms : 17602, roll : -0.01602204702794552, pitch : 0.020186109468340874, yaw : -2.1466565132141113, rollspeed : 0.010613268241286278, pitchspeed : -0.00391938304528594, yawspeed : -0.0013091671280562878}
> Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17602, q1 : -0.4772334396839142, q2 : -0.0050466363318264484, q3 : -0.011855791322886944, q4 : 0.8786820769309998, rollspeed : 0.010613268241286278, pitchspeed : -0.00391938304528594, yawspeed : -0.0013091671280562878, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
> Message from 1: HIGHRES_IMU {time_usec : 17607690, xacc : 0.20465031266212463, yacc : 0.18606188893318176, zacc : -9.777864456176758, xgyro : 0.0009166942909359932, ygyro : 0.00029493076726794243, zgyro : -0.0009300763485953212, xmag : -0.2176124006509781, ymag : 0.20348277688026428, zmag : 0.5183080434799194, abs_pressure : 86907.9921875, diff_pressure : 0.0, pressure_alt : 1276.0421142578125, temperature : 19.219999313354492, fields_updated : 7167, id : 0}
> Message from 1: ACTUATOR_CONTROL_TARGET {time_usec : 17626378, group_mlx : 0, controls : [0.013215913437306881, -0.017789743840694427, 0.00038756802678108215, 0.0, 0.0, 0.0, 0.0, -1.0]}
> Message from 1: ATTITUDE {time_boot_ms : 17622, roll : -0.01601557992398739, pitch : 0.02017391286790371, yaw : -2.1466526985168457, rollspeed : 0.009854053147137165, pitchspeed : -0.0035595863591879606, yawspeed : -0.0023903457913547754}
> Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17622, q1 : -0.4772351086139679, q2 : -0.0050427974201738834, q3 : -0.011850050650537014, q4 : 0.8786812424659729, rollspeed : 0.009854053147137165, pitchspeed : -0.0035595863591879606, yawspeed : -0.0023903457913547754, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
> Message from 1: SCALED_IMU {time_boot_ms : 17622, xacc : 16, yacc : -11, zacc : -988, xgyro : -10, ygyro : 1, zgyro : 0, xmag : -213, ymag : 218, zmag : 518, temperature : 0}
> Message from 1: SCALED_IMU2 {time_boot_ms : 17623, xacc : -15, yacc : -22, zacc : -986, xgyro : 4, ygyro : 4, zgyro : 0, xmag : 0, ymag : 0, zmag : 0, temperature : 0}
> Message from 1: SERVO_OUTPUT_RAW {time_usec : 17623436, port : 0, servo1_raw : 0, servo2_raw : 0, servo3_raw : 0, servo4_raw : 0, servo5_raw : 0, servo6_raw : 0, servo7_raw : 0, servo8_raw : 0, servo9_raw : 0, servo10_raw : 0, servo11_raw : 0, servo12_raw : 0, servo13_raw : 0, servo14_raw : 0, servo15_raw : 0, servo16_raw : 0}
> Message from 1: SERVO_OUTPUT_RAW {time_usec : 17625282, port : 1, servo1_raw : 65535, servo2_raw : 65535, servo3_raw : 65535, servo4_raw : 65535, servo5_raw : 4096, servo6_raw : 4096, servo7_raw : 65535, servo8_raw : 65535, servo9_raw : 0, servo10_raw : 0, servo11_raw : 0, servo12_raw : 0, servo13_raw : 0, servo14_raw : 0, servo15_raw : 0, servo16_raw : 0}
> Message from 1: HIGHRES_IMU {time_usec : 17627648, xacc : 0.18951965868473053, yacc : 0.15902596712112427, zacc : -9.816869735717773, xgyro : 0.00018552877008914948, ygyro : -0.0017596527468413115, zgyro : -0.001216062344610691, xmag : -0.21456921100616455, ymag : 0.20331481099128723, zmag : 0.5242573618888855, abs_pressure : 86907.0, diff_pressure : 0.0, pressure_alt : 1276.134765625, temperature : 19.219999313354492, fields_updated : 7167, id : 0}
> Message from 1: LOCAL_POSITION_NED {time_boot_ms : 17627, x : 0.0, y : 0.0, z : -0.3795027732849121, vx : 0.0208196509629488, vy : -0.009823454543948174, vz : -0.03339726850390434}
> Message from 1: ODOMETRY {time_usec : 17627648, frame_id : 18, child_frame_id : 12, x : 0.0, y : 0.0, z : -0.3795027732849121, q : [-0.4772350788116455, -0.0050383200868964195, -0.011848811991512775, 0.8786814212799072], vx : -0.0024241870269179344, vy : 0.02334427461028099, vz : -0.033083293586969376, rollspeed : 0.009519649669528008, pitchspeed : -0.0025424398481845856, yawspeed : -0.0010398122249171138, pose_covariance : [0.08489767462015152, 0.0, 0.0, 0.0, 0.0, 0.0, 0.08489584177732468, 0.0, 0.0, 0.0, 0.0, 0.05443938821554184, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], velocity_covariance : [0.0670827329158783, 0.0, 0.0, 0.0, 0.0, 0.0, 0.06707992404699326, 0.0, 0.0, 0.0, 0.0, 0.009763744659721851, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], reset_counter : 0, estimator_type : 0}
> Message from 1: ALTITUDE {time_usec : 17641953, altitude_monotonic : 1275.947021484375, altitude_amsl : 1275.947021484375, altitude_local : 0.37953153252601624, altitude_relative : 0.37953153252601624, altitude_terrain : 0.2677607238292694, bottom_clearance : 0.11177080869674683}
> Message from 1: TIMESYNC {tc1 : 0, ts1 : 17642043000}
> Message from 1: ATTITUDE {time_boot_ms : 17642, roll : -0.016015376895666122, pitch : 0.02012961357831955, yaw : -2.146639108657837, rollspeed : 0.010402746498584747, pitchspeed : -0.004715935327112675, yawspeed : -0.0019618477672338486}
> Message from 1: ATTITUDE_QUATERNION {time_boot_ms : 17642, q1 : -0.47724103927612305, q2 : -0.005023301113396883, q3 : -0.01183942798525095, q4 : 0.8786783218383789, rollspeed : 0.010402746498584747, pitchspeed : -0.004715935327112675, yawspeed : -0.0019618477672338486, repr_offset_q : [0.0, 0.0, 0.0, 0.0]}
> ```

where can I find the file "receiver.py "?
